File: packages/sdk/src/verify.ts
// Client-side verification primitives — the point of Axon is that you don't have
// to trust us, so the SDK ships the checks. No secrets leave the caller, no Axon
// endpoint sits in the trust path.
//
// Webhook signature verification lives in ./webhooks (the single canonical HMAC
// implementation). It is re-exported here so callers can import every verify
// primitive from one module.

export { verifyWebhookSignature } from "./webhooks";
export type { VerifyWebhookOptions } from "./webhooks";

// ── Proof Score verification ────────────────────────────────────────────────
// Recompute an agent's Proof Score yourself, from its public receipts — no trust
// in Axon's number. The formula below is replicated verbatim from the published
// spec (also served at /api/agents/<id>/proof-score → `formula`), so the recompute
// is byte-identical to what Axon claims.

const SCALE = 1000;
const QUALITY_WEIGHT = 0.6;
const VOLUME_WEIGHT = 0.4;
const TASKS_ANCHOR = 30;
const USDC_ANCHOR = 200;
const round = (n: number, dp = 3): number => { const f = 10 ** dp; return Math.round(n * f) / f; };
const curve = (v: number, anchor: number): number => Math.min(1, Math.log10(1 + Math.max(0, v)) / Math.log10(1 + anchor));
const provenWorkFactor = (count: number, usdc: number): number =>
  Math.min(1, 0.6 * curve(count, TASKS_ANCHOR) + 0.4 * curve(usdc, USDC_ANCHOR));

export interface VerifyProofScoreOptions {
  /** Where to fetch the proof + receipts from. Default: `https://axon-agents.com`. */
  baseUrl?: string;
  /** Inject a fetch (tests, custom agents, a different RPC-backed proxy). Default: global `fetch`. */
  fetch?: typeof fetch;
  /**
   * Re-fetch every native receipt and confirm it actually settled, instead of
   * taking the evidence list's word for it. This is the trustless step — slower
   * (one request per settled task), off by default. Cross-network items carry the
   * other network's receipt and are confirmed there.
   */
  confirmReceipts?: boolean;
}

export interface VerifyProofScoreResult {
  agentId: string;
  publishedScore: number;
  recomputedScore: number;
  scoreMatches: boolean;
  /** Settled tasks the score is computed over (the full, uncapped list). */
  evidenceCount: number;
  nativeCount: number;
  crossNetworkCount: number;
  /** null unless `confirmReceipts`; else how many native receipts re-confirmed as settled. */
  confirmedReceipts: number | null;
  /** scoreMatches AND (if confirmReceipts) every native receipt confirmed. */
  verified: boolean;
  note: string;
}

/**
 * Independently verify an agent's Proof Score. Fetches the published score and the
 * COMPLETE evidence list, recomputes the score locally from the same public
 * formula, and reports whether it matches. With `confirmReceipts`, it also
 * re-fetches every native receipt and confirms each settled — so nothing but the
 * agent's own public receipts sits in the trust path. Never trusts the score.
 */
export async function verifyProofScore(
  agentId: string,
  opts: VerifyProofScoreOptions = {},
): Promise<VerifyProofScoreResult> {
  const base = (opts.baseUrl ?? "https://axon-agents.com").replace(/\/+$/, "");
  const f = opts.fetch ?? globalThis.fetch;
  const id = encodeURIComponent(agentId);

  // The claim: published score + quality factor.
  const proofRes = await f(`${base}/api/agents/${id}/proof-score`);
  if (!proofRes.ok) throw new Error(`proof-score fetch failed: HTTP ${proofRes.status}`);
  const proof = (await proofRes.json()) as { score: number; components: { quality: { factor: number } } };

  // The evidence: the full, uncapped settled-task list.
  const evRes = await f(`${base}/api/agents/${id}/proof-score?evidence=full`);
  if (!evRes.ok) throw new Error(`evidence fetch failed: HTTP ${evRes.status}`);
  const { evidence } = (await evRes.json()) as {
    evidence: { taskId: string; network: string; verify: string | null; settledUsdc: number }[];
  };

  const native = evidence.filter((e) => e.network === "axon");
  const cross = evidence.filter((e) => e.network !== "axon");

  let confirmedReceipts: number | null = null;
  let count = evidence.length;
  let usdc = round(evidence.reduce((s, e) => s + e.settledUsdc, 0), 6);

  if (opts.confirmReceipts) {
    let ok = 0;
    let confirmedUsdc = 0;
    for (const e of native) {
      if (!e.verify) continue;
      try {
        const r = await f(`${base}${e.verify}`);
        if (!r.ok) continue;
        const receipt = (await r.json()) as { status?: string; settlement?: unknown };
        if (receipt.status === "completed" && receipt.settlement) { ok++; confirmedUsdc += e.settledUsdc; }
      } catch { /* an unreachable receipt is simply unconfirmed */ }
    }
    confirmedReceipts = ok;
    count = ok + cross.length; // cross-network items are verified on their own network
    usdc = round(confirmedUsdc + cross.reduce((s, e) => s + e.settledUsdc, 0), 6);
  }

  // Recompute, byte-identical to the published formula. Quality is taken as
  // published (reputation-derived); the proven-work half is recomputed from the
  // receipts, which is what a third party can independently confirm.
  const volumeFactor = round(provenWorkFactor(count, usdc));
  const recomputedScore = Math.round(
    round(SCALE * QUALITY_WEIGHT * proof.components.quality.factor, 2) + round(SCALE * VOLUME_WEIGHT * volumeFactor, 2),
  );

  const scoreMatches = recomputedScore === proof.score;
  const allConfirmed = confirmedReceipts === null || confirmedReceipts === native.length;
  const verified = scoreMatches && allConfirmed;

  const note = !scoreMatches
    ? `Recomputed ${recomputedScore}, but the published score is ${proof.score} — does not match.`
    : allConfirmed
      ? `Recomputed ${recomputedScore} from ${evidence.length} settled task${evidence.length !== 1 ? "s" : ""}` +
        (confirmedReceipts !== null ? ` (re-confirmed ${confirmedReceipts}/${native.length} native receipts settled)` : "") +
        "; matches the published score."
      : `Score matches, but only ${confirmedReceipts}/${native.length} native receipts confirmed settled.`;

  return {
    agentId,
    publishedScore: proof.score,
    recomputedScore,
    scoreMatches,
    evidenceCount: evidence.length,
    nativeCount: native.length,
    crossNetworkCount: cross.length,
    confirmedReceipts,
    verified,
    note,
  };
}

// ── Receipt / execution-trace verification ──────────────────────────────────
// Every Axon receipt is backed by a hash-chained execution trace (the "flight
// recorder"): each event commits to the previous event's hash, so editing,
// reordering, inserting, or deleting any NON-FINAL event breaks the chain from
// that point on. verifyReceipt fetches the public trace and RECOMPUTES the whole
// chain locally — the exact canonical-JSON + SHA-256 scheme the platform uses —
// so tamper-evidence holds without trusting Axon's own "verified" flag.
//
// Caveat, inherent to any head-less hash chain: truncating the TAIL (dropping the
// most recent events) leaves a shorter, still-internally-valid chain. verifyReceipt
// confirms the chain it is shown is intact; it cannot prove that chain is complete,
// since the trace publishes no committed length or head hash to check against.

// Deterministic JSON: keys sorted recursively, `undefined` dropped. Byte-identical
// to the scheme the trace was hashed with on write.
function canonicalStringify(value: unknown): string {
  if (value === null || typeof value !== "object") return JSON.stringify(value) ?? "null";
  if (Array.isArray(value)) return `[${value.map(canonicalStringify).join(",")}]`;
  const obj = value as Record<string, unknown>;
  const keys = Object.keys(obj).filter((k) => obj[k] !== undefined).sort();
  return `{${keys.map((k) => `${JSON.stringify(k)}:${canonicalStringify(obj[k])}`).join(",")}}`;
}

// SHA-256 hex — Web Crypto (browser + Node 18+), falling back to Node's crypto.
async function sha256hex(input: string): Promise<string> {
  if (typeof crypto !== "undefined" && crypto.subtle) {
    const digest = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(input));
    return Array.from(new Uint8Array(digest))
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("");
  }
  const { createHash } = await import("crypto");
  return createHash("sha256").update(input, "utf8").digest("hex");
}

interface PublicTraceEvent {
  seq: number;
  taskId: string | null;
  kind: string;
  fromAgent: string | null;
  toAgent: string | null;
  workflowId: string | null;
  stepIndex: number | null;
  inputHash: string | null;
  outputHash: string | null;
  model: string | null;
  inputTokens: number | null;
  outputTokens: number | null;
  costUsd: number | null;
  latencyMs: number | null;
  meta: Record<string, unknown> | null;
  prevHash: string | null;
  hash: string;
  createdAt: string;
}

export interface VerifyReceiptOptions {
  /** Where to fetch the trace from. Default: `https://axon-agents.com`. */
  baseUrl?: string;
  /** Inject a fetch (tests, custom proxy). Default: global `fetch`. */
  fetch?: typeof fetch;
}

export interface VerifyReceiptResult {
  taskId: string;
  traceId: string;
  /** Number of events in the hash chain. */
  eventCount: number;
  /** Every event's hash recomputes AND links to the previous one, with contiguous seq. */
  chainValid: boolean;
  /** seq of the first event that failed to recompute/link, or null if the chain is intact. */
  brokenAt: number | null;
  /** What the platform claims for the same chain — reported, NEVER trusted. */
  platformClaim: boolean | null;
  /** chainValid === true — the SDK's own independent verdict. */
  verified: boolean;
  note: string;
}

/**
 * Independently verify a receipt's execution trace. Fetches the public,
 * hash-chained trace for a task and recomputes every event's hash from the same
 * canonical-JSON + SHA-256 scheme used on write, checking that each links to the
 * previous (contiguous seq, matching prevHash). Nothing but the public trace sits
 * in the trust path; the platform's own `verified` flag is reported but never
 * relied on. Detects any edit, reorder, insertion, or interior deletion; cannot
 * detect tail truncation (see the module note) — `chainValid` means the shown
 * chain is intact, not provably complete.
 */
export async function verifyReceipt(
  taskId: string,
  opts: VerifyReceiptOptions = {},
): Promise<VerifyReceiptResult> {
  const base = (opts.baseUrl ?? "https://axon-agents.com").replace(/\/+$/, "");
  const f = opts.fetch ?? globalThis.fetch;
  const id = encodeURIComponent(taskId);

  const res = await f(`${base}/api/receipts/${id}/trace`);
  if (!res.ok) throw new Error(`trace fetch failed: HTTP ${res.status}`);
  const trace = (await res.json()) as {
    taskId: string;
    traceId: string;
    verified?: boolean;
    events: PublicTraceEvent[];
  };

  let prevHash: string | null = null;
  let expectedSeq = 1;
  let brokenAt: number | null = null;

  for (const e of trace.events) {
    // meta is hashed as its canonical STRING; re-canonicalize the parsed object
    // so the recompute is byte-identical to the stored bytes.
    const metaStr = e.meta == null ? null : canonicalStringify(e.meta);
    const recomputed = await sha256hex(
      canonicalStringify({
        traceId: trace.traceId,
        seq: e.seq,
        taskId: e.taskId,
        kind: e.kind,
        fromAgent: e.fromAgent,
        toAgent: e.toAgent,
        workflowId: e.workflowId,
        stepIndex: e.stepIndex,
        inputHash: e.inputHash,
        outputHash: e.outputHash,
        model: e.model,
        inputTokens: e.inputTokens,
        outputTokens: e.outputTokens,
        costUsd: e.costUsd,
        latencyMs: e.latencyMs,
        meta: metaStr,
        createdAt: e.createdAt,
        prevHash: e.prevHash,
      }),
    );
    if (e.seq !== expectedSeq || e.prevHash !== prevHash || e.hash !== recomputed) {
      brokenAt = e.seq;
      break;
    }
    prevHash = e.hash;
    expectedSeq += 1;
  }

  const chainValid = brokenAt === null && trace.events.length > 0;
  const platformClaim = typeof trace.verified === "boolean" ? trace.verified : null;
  const note = chainValid
    ? `Recomputed all ${trace.events.length} event${trace.events.length !== 1 ? "s" : ""}; the hash chain is intact.`
    : trace.events.length === 0
      ? "Trace has no events to verify."
      : `Hash chain breaks at event #${brokenAt} — the recomputed hash or link does not match.`;

  return {
    taskId: trace.taskId,
    traceId: trace.traceId,
    eventCount: trace.events.length,
    chainValid,
    brokenAt,
    platformClaim,
    verified: chainValid,
    note,
  };
}

File: packages/sdk-python/axon/verify.py
"""
Client-side verification — the point of Axon is that you can confirm claims
yourself, so the SDK ships the checks. No secrets leave the caller, no Axon
endpoint sits in the trust path.

- verify_proof_score: recompute an agent's Proof Score from its public receipts.
- verify_receipt: recompute a receipt's hash-chained execution trace.

Both are byte-identical to the server's own computation (proofScore.ts /
traceEvents.ts), so the recompute is trustworthy without trusting the number.
"""

import hashlib
import json
import math
from dataclasses import dataclass
from typing import Any, Dict, List, Optional

import requests

DEFAULT_BASE_URL = "https://axon-agents.com"


def _http_get(url: str, session: Optional[requests.Session]) -> requests.Response:
    return (session or requests).get(url, timeout=30)


# ── Proof Score ──────────────────────────────────────────────────────────────
# Replicated verbatim from the published spec (also served at
# /api/agents/<id>/proof-score -> `formula`).

_SCALE = 1000
_QUALITY_WEIGHT = 0.6
_VOLUME_WEIGHT = 0.4
_TASKS_ANCHOR = 30
_USDC_ANCHOR = 200


def _js_math_round(x: float) -> float:
    """JavaScript Math.round exactly: nearest integer, ties toward +Infinity. NOT
    floor(x + 0.5) — that diverges at the ULP just below .5 (0.49999999999999994
    rounds to 1, not 0). This is a verification primitive, so it must match the
    server (proofScore.ts) precisely — a spurious mismatch is the wrong failure."""
    fl = math.floor(x)
    diff = x - fl
    if diff > 0.5:
        return float(fl + 1)
    if diff < 0.5:
        return float(fl)
    return float(fl + 1)  # exactly .5 → round up


def _js_round(n: float, dp: int = 3) -> float:
    """JS `Math.round(n * 10^dp) / 10^dp`, matching the server's `round` helper."""
    f = 10 ** dp
    return _js_math_round(n * f) / f


def _curve(v: float, anchor: float) -> float:
    return min(1.0, math.log10(1 + max(0.0, v)) / math.log10(1 + anchor))


def _proven_work_factor(count: float, usdc: float) -> float:
    return min(1.0, 0.6 * _curve(count, _TASKS_ANCHOR) + 0.4 * _curve(usdc, _USDC_ANCHOR))


@dataclass
class ProofScoreVerification:
    agent_id: str
    published_score: int
    recomputed_score: int
    score_matches: bool
    evidence_count: int
    native_count: int
    cross_network_count: int
    confirmed_receipts: Optional[int]
    verified: bool
    note: str


def verify_proof_score(
    agent_id: str,
    base_url: str = DEFAULT_BASE_URL,
    session: Optional[requests.Session] = None,
    confirm_receipts: bool = False,
) -> ProofScoreVerification:
    """Independently verify an agent's Proof Score. Fetches the published score and
    the COMPLETE evidence list, recomputes the score locally from the same public
    formula, and reports whether it matches. With ``confirm_receipts``, it also
    re-fetches every native receipt and confirms each settled — so nothing but the
    agent's own public receipts sits in the trust path."""
    base = base_url.rstrip("/")
    aid = requests.utils.quote(agent_id, safe="")

    r = _http_get(f"{base}/api/agents/{aid}/proof-score", session)
    if not r.ok:
        raise RuntimeError(f"proof-score fetch failed: HTTP {r.status_code}")
    proof = r.json()
    quality_factor = proof["components"]["quality"]["factor"]
    published = proof["score"]

    r2 = _http_get(f"{base}/api/agents/{aid}/proof-score?evidence=full", session)
    if not r2.ok:
        raise RuntimeError(f"evidence fetch failed: HTTP {r2.status_code}")
    evidence: List[Dict[str, Any]] = (r2.json() or {}).get("evidence", [])

    native = [e for e in evidence if e.get("network") == "axon"]
    cross = [e for e in evidence if e.get("network") != "axon"]

    confirmed: Optional[int] = None
    count = len(evidence)
    usdc = _js_round(sum(e.get("settledUsdc", 0) for e in evidence), 6)

    if confirm_receipts:
        ok = 0
        confirmed_usdc = 0.0
        for e in native:
            if not e.get("verify"):
                continue
            try:
                rr = _http_get(f"{base}{e['verify']}", session)
                if not rr.ok:
                    continue
                receipt = rr.json()
                if receipt.get("status") == "completed" and receipt.get("settlement"):
                    ok += 1
                    confirmed_usdc += e.get("settledUsdc", 0)
            except Exception:  # noqa: BLE001
                pass
        confirmed = ok
        count = ok + len(cross)
        usdc = _js_round(confirmed_usdc + sum(e.get("settledUsdc", 0) for e in cross), 6)

    volume_factor = _js_round(_proven_work_factor(count, usdc))
    recomputed = int(_js_math_round(
        _js_round(_SCALE * _QUALITY_WEIGHT * quality_factor, 2)
        + _js_round(_SCALE * _VOLUME_WEIGHT * volume_factor, 2)
    ))

    score_matches = recomputed == published
    all_confirmed = confirmed is None or confirmed == len(native)
    verified = score_matches and all_confirmed
    if not score_matches:
        note = f"Recomputed {recomputed}, but the published score is {published} — does not match."
    elif all_confirmed:
        extra = f" (re-confirmed {confirmed}/{len(native)} native receipts settled)" if confirmed is not None else ""
        note = f"Recomputed {recomputed} from {len(evidence)} settled task(s){extra}; matches the published score."
    else:
        note = f"Score matches, but only {confirmed}/{len(native)} native receipts confirmed settled."

    return ProofScoreVerification(
        agent_id=agent_id,
        published_score=published,
        recomputed_score=recomputed,
        score_matches=score_matches,
        evidence_count=len(evidence),
        native_count=len(native),
        cross_network_count=len(cross),
        confirmed_receipts=confirmed,
        verified=verified,
        note=note,
    )


# ── Receipt / execution-trace verification ───────────────────────────────────
# Recompute a receipt's hash-chained trace with the exact canonical-JSON + SHA-256
# scheme it was written with. Detects any edit, reorder, insertion, or interior
# deletion. Cannot detect tail truncation (inherent to head-less hash chains) —
# `chain_valid` means the chain shown is intact, not provably complete.


def _js_number(n: Any) -> str:
    """Serialize a number byte-identically to JavaScript's JSON.stringify (the
    ECMAScript Number-to-string algorithm), so any field — including SOL amounts
    with up to 9 decimals — reproduces exactly."""
    if isinstance(n, bool):
        return "true" if n else "false"
    if isinstance(n, int):
        return str(n)
    if not isinstance(n, float):
        return "null"
    if n != n or n in (float("inf"), float("-inf")):
        return "null"
    if n == 0:
        return "0"
    sign = "-" if n < 0 else ""
    r = repr(abs(n))
    if "e" in r or "E" in r:
        mant, exp_s = r.replace("E", "e").split("e")
        exp = int(exp_s)
    else:
        mant, exp = r, 0
    int_part, frac_part = (mant.split(".") + [""])[:2]
    stripped = (int_part + frac_part).lstrip("0")
    if not stripped:
        return "0"
    s = stripped.rstrip("0")
    trailing = len(stripped) - len(s)
    k = len(s)
    n_pos = (exp - len(frac_part) + trailing) + k
    if k <= n_pos <= 21:
        out = s + "0" * (n_pos - k)
    elif 0 < n_pos <= 21:
        out = s[:n_pos] + "." + s[n_pos:]
    elif -6 < n_pos <= 0:
        out = "0." + "0" * (-n_pos) + s
    else:
        e = n_pos - 1
        out = (s[0] + ("." + s[1:] if k > 1 else "")) + "e" + ("+" if e >= 0 else "-") + str(abs(e))
    return sign + out


def _canonical(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return _js_number(value)
    if isinstance(value, str):
        return json.dumps(value, ensure_ascii=False)
    if isinstance(value, list):
        return "[" + ",".join(_canonical(v) for v in value) + "]"
    if isinstance(value, dict):
        keys = sorted(value.keys())
        return "{" + ",".join(json.dumps(k, ensure_ascii=False) + ":" + _canonical(value[k]) for k in keys) + "}"
    return "null"


@dataclass
class ReceiptVerification:
    task_id: str
    trace_id: str
    event_count: int
    chain_valid: bool
    broken_at: Optional[int]
    platform_claim: Optional[bool]
    verified: bool
    note: str


def verify_trace(trace: Dict[str, Any]) -> ReceiptVerification:
    """Recompute a fetched trace object's hash chain locally."""
    trace_id = trace.get("traceId", "")
    events: List[Dict[str, Any]] = trace.get("events") or []
    prev_hash: Optional[str] = None
    expected_seq = 1
    broken_at: Optional[int] = None
    for e in events:
        meta = e.get("meta")
        meta_str = None if meta is None else _canonical(meta)
        fields = {
            "traceId": trace_id, "seq": e.get("seq"), "taskId": e.get("taskId"),
            "kind": e.get("kind"), "fromAgent": e.get("fromAgent"), "toAgent": e.get("toAgent"),
            "workflowId": e.get("workflowId"), "stepIndex": e.get("stepIndex"),
            "inputHash": e.get("inputHash"), "outputHash": e.get("outputHash"),
            "model": e.get("model"), "inputTokens": e.get("inputTokens"),
            "outputTokens": e.get("outputTokens"), "costUsd": e.get("costUsd"),
            "latencyMs": e.get("latencyMs"), "meta": meta_str,
            "createdAt": e.get("createdAt"), "prevHash": e.get("prevHash"),
        }
        recomputed = hashlib.sha256(_canonical(fields).encode("utf-8")).hexdigest()
        if e.get("seq") != expected_seq or e.get("prevHash") != prev_hash or e.get("hash") != recomputed:
            broken_at = e.get("seq")
            break
        prev_hash = e.get("hash")
        expected_seq += 1

    chain_valid = broken_at is None and len(events) > 0
    note = (
        f"Recomputed all {len(events)} event(s); the hash chain is intact." if chain_valid
        else "Trace has no events to verify." if len(events) == 0
        else f"Hash chain breaks at event #{broken_at} — the recomputed hash or link does not match."
    )
    return ReceiptVerification(
        task_id=trace.get("taskId", ""), trace_id=trace_id, event_count=len(events),
        chain_valid=chain_valid, broken_at=broken_at,
        platform_claim=trace.get("verified") if isinstance(trace.get("verified"), bool) else None,
        verified=chain_valid, note=note,
    )


def verify_receipt(
    task_id: str,
    base_url: str = DEFAULT_BASE_URL,
    session: Optional[requests.Session] = None,
) -> ReceiptVerification:
    """Fetch a task's public execution trace and recompute the whole hash chain
    locally. The platform's own `verified` flag is reported but never relied on."""
    base = base_url.rstrip("/")
    r = _http_get(f"{base}/api/receipts/{requests.utils.quote(task_id, safe='')}/trace", session)
    if r.status_code == 404:
        raise RuntimeError(f"no execution trace found for task {task_id}")
    if not r.ok:
        raise RuntimeError(f"trace fetch failed: HTTP {r.status_code}")
    return verify_trace(r.json() or {})
